src/security/builtin_plugins/authentication/src/auth_utils.c

```diff
@@ -905,60 +903,53 @@ void free_ca_list_contents(X509Seq *ca_list)
 
 DDS_Security_ValidationResult_t get_trusted_ca_list(const char *trusted_ca_dir, X509Seq *ca_list, DDS_Security_SecurityException *ex)
 {
-  dds_return_t ret;
   ddsrt_dir_handle_t d_descr;
   struct ddsrt_dirent d_entry;
   struct ddsrt_stat status;
-  char *trusted_ca_dir_normalized;
-
-  X509 *ca_buffer_array[MAX_TRUSTED_CA];
-  unsigned ca_buffer_array_size = 0;
-  trusted_ca_dir_normalized = ddsrt_file_normalize(trusted_ca_dir);
-
-  ret = ddsrt_opendir(trusted_ca_dir_normalized, &d_descr);
-  ddsrt_free(trusted_ca_dir_normalized);
-  if (ret != DDS_RETCODE_OK || ca_buffer_array_size >= MAX_TRUSTED_CA)
+  X509 *ca_buf[MAX_TRUSTED_CA];
+  unsigned ca_cnt = 0;
+  char *tca_dir_norm = ddsrt_file_normalize(trusted_ca_dir);
+  dds_return_t ret = ddsrt_opendir(tca_dir_norm, &d_descr);
+  ddsrt_free(tca_dir_norm);
+  if (ret != DDS_RETCODE_OK)
   {
     DDS_Security_Exception_set(ex, DDS_AUTH_PLUGIN_CONTEXT, DDS_SECURITY_ERR_INVALID_TRUSTED_CA_DIR_CODE, 0, DDS_SECURITY_ERR_INVALID_TRUSTED_CA_DIR_MESSAGE);
     return DDS_SECURITY_VALIDATION_FAILED;
   }
 
-  while ((ret = ddsrt_readdir(d_descr, &d_entry)) == DDS_RETCODE_OK)
+  char *fpath, *fname;
+  X509 *ca;
+  while (ddsrt_readdir(d_descr, &d_entry) == DDS_RETCODE_OK)
   {
-    char *full_file_path;
-    ddsrt_asprintf(&full_file_path, "%s%s%s", trusted_ca_dir, ddsrt_file_sep(), d_entry.d_name);
-    if (ddsrt_stat(full_file_path, &status) == DDS_RETCODE_OK)
+    ddsrt_asprintf(&fpath, "%s%s%s", trusted_ca_dir, ddsrt_file_sep(), d_entry.d_name);
+    if (ddsrt_stat(fpath, &status) == DDS_RETCODE_OK
+      && strcmp(d_entry.d_name, ".") != 0 && strcmp(d_entry.d_name, "..") != 0
+      && (fname = ddsrt_file_normalize(fpath)) != NULL)
     {
-      if (strcmp(d_entry.d_name, ".") != 0 && strcmp(d_entry.d_name, "..") != 0)
-      {
-        char *filename = ddsrt_file_normalize(full_file_path);
-        if (filename)
-        {
-          X509 *identityCA;
-          if (load_X509_certificate_from_file(filename, &identityCA, ex) == DDS_SECURITY_VALIDATION_OK)
-          {
-            ca_buffer_array[ca_buffer_array_size] = identityCA;
-            ca_buffer_array_size++;
-          }
-          else
-            DDS_Security_Exception_reset(ex);
-          ddsrt_free(filename);
-        }
-      }
+      if (load_X509_certificate_from_file(fname, &ca, ex) == DDS_SECURITY_VALIDATION_OK)
+        ca_buf[ca_cnt++] = ca;
+      else
+        DDS_Security_Exception_reset(ex);
+      ddsrt_free(fname);
+    }
+    ddsrt_free(fpath);
+    if (ca_cnt >= MAX_TRUSTED_CA)
```

Comment:
This fails if ``MAX_TRUSTED_CA`` have been read, even though one would expect a directory with ``MAX_TRUSTED_CA`` entries to be ok. 100 is arbitrary and 99 is equally arbitrary, but it has the potential of confusing the reader. It would be better to test before adding it to the buffer, then bailing out if the maximum has already been reached.